### ISSUE TITLE
Notifications - Implementing custom group notifications

### ISSUE THREAD

**Issue by Johann-Snapp** [RocketChat/feature-requests#634]

**Comment by marceloschmidt** [comment]
I'll add a reference to RocketChat/Rocket.Chat#2135 as this is an important guide if we are going on that direction.

**Comment by KrystalMoore16** [comment]
+1 please implement this feature!

**Comment by raatti** [comment]
+2 for this, desperately needed for any good group chat to create groups and aliases to mention.

**Comment by philtweir** [comment]
(just to highlight, at least some of that bounty is time-limited - not immediately expiring, but keen to flag)

**Comment by maciek134** [comment]
Ok, so I guess @Kailash0311 decided to post a WIP PR that adds pretty much nothing to "reserve a spot"? I'm out then, I'm not going to rush anything because of a bounty... (keep in mind that this is not a trivial feature, 90% of the work is figuring out how to do it properly).

**Comment by Kailash0311** [comment]
@maciek134 I just gave it a shot by implementing a part of this feature (there is still a lot to be done and lots of refactoring has to be done) and I am not after bounty, I did it as I think this feature is really needed :slightly_smiling_face:  
Feel free to work on your own :slightly_smiling_face: .

**Comment by slyk** [comment]
Hello. We have in mid-term milestone:
1. Group to invite RocketChat/feature-requests#609
2. Mention multiple users RocketChat/feature-requests#611
3. And current about "notifications" RocketChat/feature-requests#634

They almost duplicates and solving one will solve others. I'm not good expert on GitHub, but maybe all of those requests need to be closed as duplicate of current one, to redirect all requests to one discussion and have ability to make one good solution for all?

@maciek134 any news from you? Kailash told that it's not about bounty for him. Maybe you can found some time for this feature?

P.S. added a little more bounty :)
[![Bountysource](https://camo.githubusercontent.com/8566ccd998725445bdd48a7b8402a619c7fb7243/68747470733a2f2f6170692e626f756e7479736f757263652e636f6d2f62616467652f69737375653f69737375655f69643d3330373533333530)](https://www.bountysource.com/issues/30753350-notifications-implementing-custom-group-notifications?utm_source=30753350&utm_medium=shield&utm_campaign=ISSUE_BADGE)

**Comment by maciek134** [comment]
I'll take another stab at it :)

**Comment by SeaLife** [comment]
Hopefully this can be used per-channel instead of per-instance.

In a large organization with like 8000 (or nearly 20k, like in RocketChat/Rocket.Chat#1315) people using rocket.chat, global groups/roles to be mention will be useless. But havin a channel (like Discord has 'servers') where you can mention the @developers or @templating would help alot.

So there must be a per-channel custom notification system?

**Comment by maciek134** [comment]
I've got some free time on my hands so I'll try to have something before the end of the week.

@SeaLife yeah, that's what I was thinking - but global groups too, so you can be in a QA group but if someone mentions `@qa` in a channel you will only get notified if you already belong to it.

**Comment by SeaLife** [comment]
Did you end up with something? @maciek134 :)

**Comment by maciek134** [comment]
A lot of bad feelings for meteor ;p Seriously though, I've made some
progress - figured out the notification part, now I need to finish settings
and I'll have something to show.

**Comment by yuritoledo** [comment]
+1

**Comment by Crayder** [comment]
This would also be really awesome with an addition of the option to only notify people marked active.

Even further, you could add online-detection and these custom tags to the bot's api. 

This would help for customer notification systems for business-class users like the company I work for a lot!

**Comment by J0kerFace** [comment]
This is our mostly missing function at the moment.

**Comment by maciek134** [comment]
I have good and bad news - bad news is that our company no longer uses Rocket internally so I need to work on this in my free time only. Good news is that this is my top priority now, so you can expect this to be done shortly ;p

**Comment by maciek134** [comment]
Progress update (or more like a teaser):
![image](https://user-images.githubusercontent.com/1146309/69277537-baa24d00-0be0-11ea-9181-1b61dba44293.png)
Got right into zone after work today and finished most of the work, should have a PR ready tomorrow/friday.

**Comment by J0kerFace** [comment]
looks really cool! we are looking forward to test it out

**Comment by SeaLife** [comment]
Hopefully this can be set independentlly in each room to not have the sysadmins create notifications groups in a 8000+ employeed company :D 

But this looks really nice.

**Comment by Wirone** [comment]
@SeaLife personally I don't think notification groups should be defined per channel, or at least there should possibility to choose type of such group. We would like to make groups per team, so on every channel team could be mentioned and notified in the same way (defined once, used in multiple rooms). But on the other hand I see scenarios, where group per room could be useful - for example `maintainers` for per-project rooms.. @maciek134 what do you think?

**Comment by maciek134** [comment]
I tried to think about everything mentioned here, so I already have a solution for this (at work atm so can't post a screenshot) - when creating/editing group you have the following options:
- name
- description
- whether the group also notifies a specific role
- whether the group also notifies a specific group (so in combination with a few other options you can make a transparent copy that notifies offline users too)
- whether users can join this group themselves
- whether the group notifies offline people (like `@all` vs `@here`)
- whether the group notifies people outside a channel
- whether the group works in specific channels only
I guess there also should be an option to make it so users are also assigned per-room but I'd make it further down the line since this is a pretty specific scenario that afaik nobody asked for yet - should be relatively easy though.

**Comment by SeaLife** [comment]
I agree with this in general, but we'r using RC with over 8000 employees. Having custom groups per Channel and creatable by the channel Admin would be really nice to not bother the Admins with notifications groups. :)

**Comment by Crayder** [comment]
>whether the group notifies offline people (like `@all` vs `@here`)

Thank you!

**Comment by Wirone** [comment]
@SeaLife but with granular permissions in RocketChat it could be possible to create role for managing notification groups and allow users to do it, so Admins won't be bothered that much. I don't think notification groups configuration should go to channel's settings panel.

**Comment by SeaLife** [comment]
@Wirone okay, we should probably evaluate this first...
Its probably not that good if everyone can create global notification groups...
But we'll see :)

Im looking forward to this :)

**Comment by maciek134** [comment]
This should only take a day or two more, I re-did the admin side of things since it was a bit crammed in a side panel when editing groups.

**Comment by J0kerFace** [comment]
What's the status here?

**Comment by maciek134** [comment]
Got overwhelmed with other work, I hope to have this finished by the end of the week.

EDIT: nothing came of the plans, but I'm taking the rest of the month off Thursday, so I will have plenty of time ;p

**Comment by sdsznsk** [comment]
@maciek134 Any updates on this? It would help a lot

**Comment by maciek134** [comment]
@Teqashi I'm almost done, just didn't have time to wrap it up and send a PR. I'll try to do it in the evening.

**Comment by fdellwing** [comment]
We just had a meeting, discussing to supersede RC because not custom mentions are possible. Really looking forward into testing this!

**Comment by maciek134** [comment]
@fdellwing I'm pushing it today. Btw what were you thinking of replacing RC with? ~I'm not aware of any that implement custom mentions atm.~ Oh, Slack has it as "user groups"

**Comment by fdellwing** [comment]
That's correct, we looked at Slack and Teams. But we would like to further use RC, because we really like the `discussions`.

**Comment by J0kerFace** [comment]
HipChat have it too. But that's been discontinued.

**Comment by compuguy** [comment]
@maciek134 When you get a chance, can you please link to your pull request when you open it?

**Comment by maciek134** [comment]
GitHub links PRs with issue mentions automatically ;p It's pretty much done, but user self-assignment is missing along with a few translations.

**Comment by compuguy** [comment]
I didn't know that @maciek134. I understand that this is a WIP. It will be worth the wait though.

**Comment by thesystemninjaneer** [comment]
Not sure if this is the right Issue to request this but can the functionality to message using `@owner` and `@leader` of public and private channels be added as well?

We have CI/CD pipelines automatically notifying channels of pass/fail build status and the owner/leader of the channel cares about the build status the most. Support for those native roles saves us a few extra steps of having to setup a group and and/remove members over time as employees change their roles or move on in their career while the channel Leader/Owner consistently remains up to date.

**Comment by Crayder** [comment]
@thesystemninjaneer That's basically what this whole issue is suggesting, custom groups and custom notification filtering for those groups. With this implementation you'll be able to make those groups and assign people to them like normal except now you will be able to only notify active people and stuff. For being the owner of this particular group you'd just create a group for the channel in question called "owner" and assign yourself to it.

**Comment by thesystemninjaneer** [comment]
@Crayder I’m tracking the Group approach coming with this the feature in this isse but was asking for a way to notify the person assigned a particular channel role (such as owner or leader) without the extra steps of creating and maintaining a custom group. Is that something easily achievable?

**Comment by maciek134** [comment]
@thesystemninjaneer easily achievable - yes, but I don't think this is in the scope of this issue / PR.

**Comment by Shaverdoff** [comment]
any progress of that? or this develop-issues need some donate?

**Comment by maciek134** [comment]
@Shaverdoff I may finally have some time to finish this one, I'll try to push fixes this week.

**Comment by sviat9440** [comment]
Any news ?

**Comment by julianeisel** [comment]
Let me second this! Such a feature would be very helpful for us at Blender ([blender.chat](https://blender.chat/)), and allow us to move more conversations to public channels.

There's lots of activity in our public channels, if I want to notify a specific team (say the core UI-team) it's a hassle to do in public channels. Hence some teams end up using lower-traffic, private team channels more, which is against our own transparency-principles. 
We use read-only channels in some instances, but they further fragment public communication which is an issue already.

**Comment by sviat9440** [comment]
@maciek134 are you have problem? 2021 still waiting (
This is very important feature for us...

**Comment by vinogradovia** [comment]
Any news ?

**Comment by fdellwing** [comment]
https://github.com/RocketChat/Rocket.Chat/pull/16311#issuecomment-825389926 ff

We are now switching to teams. They at least are not putting essential features behind a paywall. Without the option to buy it separately.

**Comment by vinogradovia** [comment]
thx,  strange that I did not find this functionality in the price (licence comparison)

**Comment by fdellwing** [comment]
https://docs.rocket.chat/guides/user-guides/rooms/teams/teams-enterprise-edition

It's not even included in the middle price option, only in the highest price that they do not even communicate openly.
